packages/emnapi/src/emnapi.c

```diff
@@ -154,69 +156,108 @@ emnapi_get_emscripten_version(napi_env env,
 
 #ifdef __EMSCRIPTEN_PTHREADS__
 
+#include <string.h>
+#include "threadpool.c"
+
+typedef void (*_emnapi_call_into_module_callback)(napi_env env, void* args);
+extern void _emnapi_call_into_module(napi_env env, _emnapi_call_into_module_callback callback, void* args);
+
 struct napi_async_work__ {
   napi_env env;
+  void* data;
   napi_async_execute_callback execute;
   napi_async_complete_callback complete;
-  void* data;
-  pthread_t tid;
+  uv_work_t work_req_;
 };
 
-typedef struct worker_count {
-  int unused;
-  int running;
-} worker_count;
-
-// extern void _emnapi_create_async_work_js(napi_async_work work);
-// extern void _emnapi_delete_async_work_js(napi_async_work work);
-extern void _emnapi_queue_async_work_js(napi_async_work work);
-extern void _emnapi_on_execute_async_work_js(napi_async_work work);
-extern int _emnapi_get_worker_count_js(worker_count* count);
-
-static napi_async_work _emnapi_async_work_init(
+static napi_async_work async_work_init(
   napi_env env,
   napi_async_execute_callback execute,
   napi_async_complete_callback complete,
   void* data
 ) {
   napi_async_work work = (napi_async_work)malloc(sizeof(struct napi_async_work__));
   if (work == NULL) return NULL;
+  memset(work, 0, sizeof(struct napi_async_work__));
```

Comment:
Nit: you can just use calloc instead. It's both shorter and, at least on native platforms, more efficient, so it's just a good habit.